openff/system/components/foyer.py

```diff
@@ -0,0 +1,152 @@
+from copy import copy
+from typing import Any, Dict, Set, Tuple
+
+import parmed as pmd
+from foyer import Forcefield  # type: ignore
+from openff.toolkit.topology import Topology
+
+from openff.system import unit as u
+from openff.system.components.potentials import Potential, PotentialHandler
+from openff.system.components.system import System
+from openff.system.exceptions import MissingParametersError
+
+
+def _copy_params(
+    params: Dict[str, float], *drop_keys: str, param_units: Dict = None
+) -> Dict:
+    params_copy = copy(params)
+    for drop_key in drop_keys:
+        params_copy.pop(drop_key, None)
+    if param_units:
+        for unit_item, units in param_units.items():
+            params_copy[unit_item] = params_copy[unit_item] * units
+    return params_copy
+
+
+def _get_openmm_force_gen(ff: Forcefield, gen_type: type) -> Any:
+    return list(filter(lambda x: isinstance(x, gen_type), ff.getGenerators())).pop()
+
+
+def _topology_from_parmed(structure: pmd.Structure) -> Topology:
+    """ToDo: Create a openFF Topology from a parmed structure"""
+
+
+def from_foyer(structure: pmd.Structure, ff: Forcefield, **kwargs) -> System:
+    """Create an openFF system object from a parmed structure by applying a foyer Forcefield"""
+    system = System()
+    system.topology = _topology_from_parmed(structure)
+    foyer_atom_handler = FoyerAtomTypes()
+    foyer_bond_handler = FoyerBondHandler()
+    system.handlers["FoyerAtomHandler"] = foyer_atom_handler
+    system.handlers["FoyerBondHandler"] = foyer_bond_handler
+
+    return system
+
+
+class FoyerAtomTypes(PotentialHandler):
+
+    name: str = "Atoms"
+    expression: str = "4*epsilon*((sigma/r)**12-(sigma/r)**6)"
+    independent_variables: Set[str] = {"r"}
+    slot_map: Dict[int, str] = dict()  # type: ignore
+    potentials: Dict[str, Potential] = dict()
+
+    def store_matches(
+        self,
+        forcefield: Forcefield,
+        structure: pmd.Structure,
+    ) -> None:
```

Comment:
Would it be possible to modify this function in a way that bypasses a reliance on ParmEd? I know it is currently used as an intermediary data store, but I think generating matches directly from an `mb.Compound` would be more consistent with the scope of this function. It may also be worth exploring a larger refactor that allows Foyer to operate directly on compounds with less reliance on ParmEd and OpenMM.

```suggestion
    def store_matches(
        self,
        forcefield: Forcefield,
        compound: mb.Compound,
    ) -> None:
```

Reply:
With https://github.com/mosdef-hub/foyer/pull/358, I think we can either take in an untyped gmso `Topology` if so desired. This step needs some extension to support a `mBuild` compound IMO. https://github.com/mosdef-hub/foyer/blob/82ebfca79044af9f2fb449b91941fdc8071e6c5d/foyer/forcefield.py#L583